### ISSUE TITLE
Updates broken from 4.11.0 to 4.13.3

### ISSUE THREAD

**Issue by r7l** [matomo-org/tag-manager#598]
I am not sure if this should be here or in Matomo repo. But the core of the issue come from this repo.

When trying to update a DB from Matomo 4.11.0 to 4.13.3 it fails with an error. The error shown on page is barely saying much as it just complains about a missing class. But when using the `console` it shows it in more detail:

```

    *** Update ***

    Database Upgrade Required

    Your Matomo database is out-of-date, and must be upgraded before you can continue.

    Matomo database will be upgraded from version 4.11.0 to the new version 4.13.3.

    The following plugins will be updated: TagManager.

ERROR [2023-02-02 11:51:43] 70  Uncaught exception: Error: The class \Piwik\Plugins\TagManager\Updates_newtagparametermigrator was not found in /srv/http/public/htdocs/plugins/TagManager/Updates/NewTagParameterMigrator.php in /srv/http/public/htdocs/core/Updater.php:247
Stack trace:
#0 /srv/http/public/htdocs/plugins/CoreUpdater/Commands/Update.php(355): Piwik\Updater->getSqlQueriesToExecute()
#1 /srv/http/public/htdocs/plugins/CoreUpdater/Commands/Update.php(170): Piwik\Plugins\CoreUpdater\Commands\Update->getMigrationQueriesToExecute()
#2 /srv/http/public/htdocs/plugins/CoreUpdater/Commands/Update.php(162): Piwik\Plugins\CoreUpdater\Commands\Update->doDryRun()
#3 /srv/http/public/htdocs/plugins/CoreUpdater/Commands/Update.php(65): Piwik\Plugins\CoreUpdater\Commands\Update->makeUpdate()
#4 /srv/http/public/htdocs/vendor/symfony/console/Symfony/Component/Console/Command/Command.php(257): Piwik\Plugins\CoreUpdater\Commands\Update->execute()
#5 /srv/http/public/htdocs/vendor/symfony/console/Symfony/Component/Console/Application.php(874): Symfony\Component\Console\Command\Command->run()
#6 /srv/http/public/htdocs/vendor/symfony/console/Symfony/Component/Console/Application.php(195): Symfony\Component\Console\Application->doRunCommand()
#7 /srv/http/public/htdocs/core/Console.php(108): Symfony\Component\Console\Application->doRun()
#8 [internal function]: Piwik\Console->originDoRun()
#9 /srv/http/public/htdocs/core/Console.php(147): call_user_func()
#10 /srv/http/public/htdocs/core/Access.php(670): Piwik\Console->Piwik\{closure}()
#11 /srv/http/public/htdocs/core/Console.php(148): Piwik\Access::doAsSuperUser()
#12 /srv/http/public/htdocs/core/Console.php(87): Piwik\Console->doRunImpl()
#13 /srv/http/public/htdocs/vendor/symfony/console/Symfony/Component/Console/Application.php(126): Piwik\Console->doRun()
#14 /srv/http/public/htdocs/console(32): Symfony\Component\Console\Application->run()
#15 {main}
Uncaught exception in /srv/http/public/htdocs/core/Updater.php line 247:
The class \Piwik\Plugins\TagManager\Updates_newtagparametermigrator was not found in /srv/http/public/htdocs/plugins/TagManager/Updates/NewTagParameterMigrator.php
```

I've tracked it down within the `Updater.php` file and from what i can see, it seems as if there should be files that have been deleted.

It loops through the possible files and expects these strings:

- 4.12.0-b1
- 4.12.0-b2
- 4.12.0-b3
- 4.12.0-b4
- NewTagParameterMigrator

Non of these versions are shipped anymore. Currently there are only these strings possible:

- 4.12.3-b1
- 5.0.0-b1.php

So it will just take the `NewTagParameterMigrator.php` file which breaks the entire thing.

It took me a while to figure this out. The 4.11.0 was released later in 2022. It should have been updated early, i can see that.
But i was handed this setup to update it now. It's still just a couple months old version and there isn't any information about these kind of breaking changes when updating a Matomo version that wasn't even a major release. The site is completely unusable with just throwing this error. 

There should at least be a better output on page when doing updates like this saying that Updates should go from 4.11 to 4.12 to 4.13 or something. Or maybe you just keep the older files around longer.

**Comment by r7l** [comment]
Just checked other versions. It seems as if you only keep the most recent version. I have this error with any of the version i am downloading. Is this how it is supposed to be or am i doing it wrong? People are actually required to download all the versions and cycle through them when missing a couple of updates?

I am also adding the screenshot of the original error shown to me:

![Screenshot from 2023-02-02 13-40-51](https://user-images.githubusercontent.com/30982145/216327996-35e8446b-f75d-4e0c-beee-d635ca4deb0b.png)

**Comment by AltamashShaikh** [comment]
@r7l Can you please list the steps you used for upgrading the Matomo version ?
I tried replicating on my local and can see it picks only the 2 files not the `New*` files
![Screenshot from 2023-02-02 18-49-09](https://user-images.githubusercontent.com/14902277/216341302-aff8fbaa-e0a8-4866-a216-31c464186ec5.png)

**Comment by r7l** [comment]
@AltamashShaikh Thanks for your quick response.

The general setup is exactly the same to another running (up to date) Matomo setup i have for another but unreleated project. 

I've got that DB dump of the Matomo install without the Matomo files. Then i've downloaded the most recent release of Matomo and put the files in place. Placing the dump in MariaDB and editing the config accordingly. So after that, i was just opening up Matomo in browser. The first error was about the missing `salt` which i provided using `?i_am_super_user=...`. So once i had this, it was showing me the error in question.

I've even tried without the browser. Having the dump freshly in my MariaDB and running `./console core:update` runs into this error straight away.

**Comment by AltamashShaikh** [comment]
@r7l what is the output of this query?`Select * from matomo_option where option_name= 'version_TagManager'`

**Comment by r7l** [comment]
For some reason it's not in there. But there are other plugins listed like this:

![Screenshot from 2023-02-02 15-24-45](https://user-images.githubusercontent.com/30982145/216350911-9861b65e-b3ed-49d6-ad00-e1fbc69d6b73.png)

**Comment by r7l** [comment]
Let me add again that i was just handed this dump. I don't have access to the original setup and i don't really know why it isn't in there. It didn't even occur to me that it would be possible to break Matomo like this.

I've also found their update paths in that table: `MatomoUpdateHistory | 4.10.0,4.9.1,4.2.1,4.1.1,3.14.1`

**Comment by r7l** [comment]
Listing the pluging with `console plugin:list` shows the Tag Manager to be enabled:

```
+----------------------+-------------------+---------------+
| Plugin               | Core or optional? | Status        |
+----------------------+-------------------+---------------+
| Diagnostics          | Core              | Activated     |
| Login                | Core              | Activated     |
| CoreAdminHome        | Core              | Activated     |
| UsersManager         | Core              | Activated     |
| SitesManager         | Core              | Activated     |
| Installation         | Core              | Activated     |
| Monolog              | Core              | Activated     |
| Intl                 | Core              | Activated     |
| CoreVue              | Core              | Activated     |
| CorePluginsAdmin     | Core              | Activated     |
| CoreHome             | Core              | Activated     |
| WebsiteMeasurable    | Core              | Activated     |
| IntranetMeasurable   | Core              | Activated     |
| CoreVisualizations   | Core              | Activated     |
| Proxy                | Core              | Activated     |
| API                  | Core              | Activated     |
| Widgetize            | Core              | Activated     |
| Transitions          | Core              | Activated     |
| LanguagesManager     | Core              | Activated     |
| Actions              | Core              | Activated     |
| Dashboard            | Core              | Activated     |
| MultiSites           | Core              | Activated     |
| Referrers            | Core              | Activated     |
| UserLanguage         | Core              | Activated     |
| DevicesDetection     | Core              | Activated     |
| Goals                | Core              | Activated     |
| Ecommerce            | Core              | Activated     |
| SEO                  | Core              | Activated     |
| Events               | Core              | Activated     |
| UserCountry          | Core              | Activated     |
| GeoIp2               | Core              | Activated     |
| VisitsSummary        | Core              | Activated     |
| VisitFrequency       | Core              | Activated     |
| VisitTime            | Core              | Activated     |
| VisitorInterest      | Core              | Activated     |
| RssWidget            | Core              | Activated     |
| Feedback             | Core              | Activated     |
| TwoFactorAuth        | Core              | Activated     |
| CoreUpdater          | Core              | Activated     |
| CoreConsole          | Core              | Activated     |
| ScheduledReports     | Core              | Activated     |
| UserCountryMap       | Core              | Activated     |
| Live                 | Core              | Activated     |
| PrivacyManager       | Core              | Activated     |
| ImageGraph           | Core              | Activated     |
| Annotations          | Core              | Activated     |
| MobileMessaging      | Core              | Activated     |
| Overlay              | Core              | Activated     |
| SegmentEditor        | Core              | Activated     |
| Insights             | Core              | Activated     |
| Morpheus             | Core              | Activated     |
| Contents             | Core              | Activated     |
| BulkTracking         | Core              | Activated     |
| Resolution           | Core              | Activated     |
| DevicePlugins        | Core              | Activated     |
| Heartbeat            | Core              | Activated     |
| ProfessionalServices | Core              | Activated     |
| UserId               | Core              | Activated     |
| CustomJsTracker      | Core              | Activated     |
| PagePerformance      | Core              | Activated     |
| CustomDimensions     | Core              | Activated     |
| Marketplace          | Core              | Activated     |
| Tour                 | Core              | Activated     |
| TagManager           | Core              | Activated     |
| Provider             | Optional          | Not activated |
+----------------------+-------------------+---------------+
```

But i've just checked the SQL dump and there is no table from tag manager in there. Adding that missing line to `option` table will just add other SQL errors looking for tables of Tag Manager:

```
Base table or view not found: 1146 Table 'public.tagmanager_container' doesn't exist
```

So something is wrong with this dump. Thanks allot for your help!

**Comment by r7l** [comment]
@AltamashShaikh I've figured the issue. I've copied over the `config.inc.php` as the setup it was identical in my testing environment to the other install i have for testing purposes. It was basically containing logins to MySQL und Redis and i wasn't aware Matomo changed it to write enabled Plugins there. 

For some reason the Dump had the plugin not enabled while my other install had. Removing the list of Plugins from the config fixed the issue.

Sorry for causing trouble like this. I would not have expected a software to write plugin states into a file when having a database at hand but i am aware of it now. Should have checked that config file.

Thanks again for pointing me into the right direction!
